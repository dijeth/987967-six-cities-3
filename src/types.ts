export type id = string;
export type roomType = `apartment` | `room` | `house` | `hotel`;
export type coord = [number, number];