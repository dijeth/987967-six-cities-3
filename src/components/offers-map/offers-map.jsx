import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import leaflet from 'leaflet';
import {isEqualCoords} from '../../util.js';
import {connect} from 'react-redux';
import {getActiveOfferCoord} from '../../reducers/app/selectors.js';

const ICON_SIZE = [27, 39];

const ICON = leaflet.icon({
  iconUrl: `img/pin.svg`,
  iconSize: ICON_SIZE
});

const ICON_ACTIVE = leaflet.icon({
  iconUrl: `img/pin-active.svg`,
  iconSize: ICON_SIZE
});

class OffersMap extends PureComponent {
  constructor(props) {
    super(props);

    this._centerCoord = null;
    this._activeCoord = null;
    this._offersCoord = [];
    this._activeLayer = null;
    this._offerLayers = [];

    this._mapWrapper = React.createRef();
  }

  componentDidMount() {
    const {centerCoord, offersCoord, activeCoord, zoom} = this.props;

    this._mapWrapper.current.style.height = `100%`;

    this._map = leaflet.map(this._mapWrapper.current, {
      center: centerCoord,
      zoom,
      zoomControl: false,
      marker: true
    });

    leaflet
      .tileLayer(
          `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
            attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
          })
      .addTo(this._map);

    this._setView(centerCoord, zoom);

    this._centerCoord = centerCoord;
    this._updateOffers(offersCoord);
    this._updateActiveCoord(activeCoord);
  }

  componentDidUpdate() {
    const {centerCoord, offersCoord, activeCoord, zoom} = this.props;

    if (this._shouldOfferMarkersBeChanged(centerCoord, offersCoord)) {
      this._centerCoord = centerCoord;

      this._setView(centerCoord, zoom);
      this._updateOffers(offersCoord);
    }

    this._updateActiveCoord(activeCoord);
  }

  componentWillUnmount() {
    this._map.remove();
  }

  _shouldOfferMarkersBeChanged(centerCoord, offersCoord) {
    if (!isEqualCoords(centerCoord, this._centerCoord)) {
      return true;
    }

    if (offersCoord.length !== this._offersCoord.length) {
      return true;
    }

    if (offersCoord.filter((it, i) => !isEqualCoords(it, this._offersCoord[i])).length) {
      return true;
    }

    return false;
  }

  _setView(coords, zoom) {
    this._map.setView(coords, zoom);
  }

  _addOffers(offersCoord) {
    this._offersCoord = offersCoord;
    return offersCoord.map((coord) => leaflet.marker(coord, {icon: ICON}).addTo(this._map));
  }

  _updateOffers(offersCoord) {
    this._offerLayers.forEach((it) => {
      this._map.removeLayer(it);
    });

    this._offerLayers = null;
    this._offerLayers = this._addOffers(offersCoord);
  }

  _updateActiveCoord(coord) {
    if (this._activeLayer !== null) {
      this._map.removeLayer(this._activeLayer);
      this._activeLayer = null;
    }

    if (coord) {
      this._activeLayer = leaflet.marker(coord, {icon: ICON_ACTIVE}).addTo(this._map);
    }
  }

  render() {
    return <div id='map' ref={this._mapWrapper}></div>;
  }
}

const mapStateToProps = (state) => ({
  activeCoord: getActiveOfferCoord(state)
});

OffersMap.propTypes = {
  centerCoord: PropTypes.arrayOf(PropTypes.number).isRequired,
  offersCoord: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
  activeCoord: PropTypes.arrayOf(PropTypes.number),
  zoom: PropTypes.number
};

export {OffersMap};
export default connect(mapStateToProps)(OffersMap);
