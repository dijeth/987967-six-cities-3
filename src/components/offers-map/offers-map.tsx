import * as React from 'react';
import * as leaflet from 'leaflet';
import {isEqualCoords} from '../../util.js';
import {connect} from 'react-redux';
import {getActiveOfferCoord} from '../../reducers/app/selectors.js';
import {coord} from '../../types';

const ICON_SIZE: [number, number] = [27, 39];

const ICON = leaflet.icon({
  iconUrl: `img/pin.svg`,
  iconSize: ICON_SIZE
});

const ICON_ACTIVE = leaflet.icon({
  iconUrl: `img/pin-active.svg`,
  iconSize: ICON_SIZE
});

const LAYER_URL = `https://{s}.basemaps.cartocdn.com/rastertiles/voyager_labels_under/{z}/{x}/{y}{@2x}.png`;

type Props = {
  centerCoord: coord;
  offersCoord: Array<coord>;
  activeCoord: coord;
  zoom: number;
};

class OffersMap extends React.PureComponent<Props> {
  private mapRef: React.RefObject<HTMLDivElement>;
  private map: leaflet.Map;
  private layerGroup: leaflet.LayerGroup;
  private activeLayer: leaflet.Marker;

  private offersCoord: Array<coord>;
  private activeCoord: coord;
  private centerCoord: coord;

  constructor(props) {
    super(props);
    this.mapRef = React.createRef();
    this.map = null;

    this.layerGroup = null;
    this.activeLayer = null;

    this.offersCoord = null;
    this.activeCoord = null;
    this.centerCoord = null;
  }

  updateActiveLayer() {
    if (
      JSON.stringify(this.activeCoord) ===
      JSON.stringify(this.props.activeCoord)
    ) {
      return;
    }

    this.activeCoord = this.props.activeCoord;

    if (this.activeLayer !== null) {
      this.activeLayer.remove();
      this.activeLayer = null;
    }

    if (this.activeCoord !== null) {
      this.activeLayer = leaflet
        .marker(this.activeCoord, { icon: ICON_ACTIVE })
        .addTo(this.map);
    }
  }

  updateLayerGroup() {
    if (
      JSON.stringify(this.offersCoord) ===
      JSON.stringify(this.props.offersCoord)
    ) {
      return;
    }

    this.offersCoord = this.props.offersCoord;
    this.layerGroup.clearLayers();

    this.offersCoord.forEach(it => {
      leaflet.marker(it, { icon: ICON }).addTo(this.layerGroup);
    });
  }

  updateView() {
    if (
      JSON.stringify(this.centerCoord) ===
      JSON.stringify(this.props.centerCoord)
    ) {
      return;
    };

    this.centerCoord = this.props.centerCoord;
    this.map.setView(this.centerCoord, this.props.zoom);
  }

  componentDidUpdate() {
    this.updateActiveLayer();
    this.updateLayerGroup();
    this.updateView();
  }

  componentDidMount() {
    this.map = leaflet.map(this.mapRef.current.id, {
      zoomControl: false,
      scrollWheelZoom: false
    });

    leaflet.tileLayer(LAYER_URL).addTo(this.map);
    
    this.layerGroup = leaflet.layerGroup().addTo(this.map);

    this.updateActiveLayer();
    this.updateLayerGroup();
    this.updateView();
  }

  render() {
    return (
      <div
        ref={this.mapRef}
        id="map"
        style={{ width: `100%`, height: `100%` }}
      ></div>
    );
  }
}

const mapStateToProps = (state) => ({
  activeCoord: getActiveOfferCoord(state)
});

export {OffersMap};
export default connect(mapStateToProps)(OffersMap);
