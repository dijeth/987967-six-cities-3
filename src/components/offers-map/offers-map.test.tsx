import * as React from 'react';
import {mount} from 'enzyme';
import {OffersMap} from './offers-map';
import {coord} from '../../types';

const props = {
  zoom: 16,
  activeCoord: [52.377304, 4.903076] as coord,
  centerCoord: [52.38333, 4.9] as coord,
  offersCoord: [
    [52.359385, 4.879898],
    [52.353995, 4.911789],
    [52.377303, 4.903075]
  ] as Array<coord>,
};

it(`<OffersMap /> should be render correctly with activeCoord`, () => {
  const div = document.createElement(`div`);
  document.body.appendChild(div);

  const map = mount(<OffersMap {...props} />, {attachTo: div});
  expect(map.getDOMNode()).toMatchSnapshot();
});


