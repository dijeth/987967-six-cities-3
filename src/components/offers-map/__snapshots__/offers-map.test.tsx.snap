// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OffersMap /> should be render correctly with activeCoord 1`] = `
<div
  class="leaflet-container leaflet-grab leaflet-touch-drag"
  id="map"
  style="width: 100%; height: 100%; position: relative;"
  tabindex="0"
>
  <div
    class="leaflet-pane leaflet-map-pane"
    style="left: 0px; top: 0px;"
  >
    <div
      class="leaflet-pane leaflet-tile-pane"
    >
      <div
        class="leaflet-layer "
        style="z-index: 1;"
      >
        <div
          class="leaflet-tile-container leaflet-zoom-animated"
          style="z-index: 18; left: 0px; top: 0px;"
        >
          <img
            alt=""
            class="leaflet-tile"
            role="presentation"
            src="https://c.basemaps.cartocdn.com/rastertiles/voyager_labels_under/16/33660/21533{@2x}.png"
            style="width: 256px; height: 256px; left: -5px; top: -178px;"
          />
        </div>
      </div>
    </div>
    <div
      class="leaflet-pane leaflet-shadow-pane"
    />
    <div
      class="leaflet-pane leaflet-overlay-pane"
    />
    <div
      class="leaflet-pane leaflet-marker-pane"
    >
      <img
        alt=""
        class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
        src="img/pin.svg"
        style="margin-left: -13.5px; margin-top: -19.5px; width: 27px; height: 39px; left: -937px; top: 1828px; z-index: 1828;"
        tabindex="0"
      />
      <img
        alt=""
        class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
        src="img/pin.svg"
        style="margin-left: -13.5px; margin-top: -19.5px; width: 27px; height: 39px; left: 549px; top: 2239px; z-index: 2239;"
        tabindex="0"
      />
      <img
        alt=""
        class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
        src="img/pin.svg"
        style="margin-left: -13.5px; margin-top: -19.5px; width: 27px; height: 39px; left: 143px; top: 460px; z-index: 460;"
        tabindex="0"
      />
      <img
        alt=""
        class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
        src="img/pin-active.svg"
        style="margin-left: -13.5px; margin-top: -19.5px; width: 27px; height: 39px; left: 143px; top: 460px; z-index: 460;"
        tabindex="0"
      />
    </div>
    <div
      class="leaflet-pane leaflet-tooltip-pane"
    />
    <div
      class="leaflet-pane leaflet-popup-pane"
    />
  </div>
  <div
    class="leaflet-control-container"
  >
    <div
      class="leaflet-top leaflet-left"
    />
    <div
      class="leaflet-top leaflet-right"
    />
    <div
      class="leaflet-bottom leaflet-left"
    />
    <div
      class="leaflet-bottom leaflet-right"
    >
      <div
        class="leaflet-control-attribution leaflet-control"
      >
        <a
          href="https://leafletjs.com"
          title="A JS library for interactive maps"
        >
          Leaflet
        </a>
      </div>
    </div>
  </div>
</div>
`;
